plugin/completion.py

```diff
@@ -41,81 +41,68 @@
-        if not session:
-            return
 
-        client = client_from_session(session)
-        if not client:
-            return
+class LspCompleteInsertTextCommand(LspCompleteCommand):
 
-        completion_provider = session.get_capability('completionProvider')
-        has_resolve_provider = completion_provider and completion_provider.get('resolveProvider', False)
-        if has_resolve_provider:
-            client.send_request(Request.resolveCompletionItem(item), self.handle_resolve_response)
+    def run(self, edit: sublime.Edit, **item: Any) -> None:
+        insert_text = item.get("insertText") or item["label"]
+        if item.get("insertTextFormat", InsertTextFormat.PlainText) == InsertTextFormat.Snippet:
+            self.view.run_command("insert_snippet", {"contents": insert_text})
+        else:
+            self.view.run_command("insert", {"characters": insert_text})
+        self.handle_additional_edits(item)
 
-    def handle_resolve_response(self, response: Optional[dict]) -> None:
-        if response:
-            additional_edits = response.get('additionalTextEdits')
-            if additional_edits:
-                self.apply_additional_edits(additional_edits)
 
-    def apply_additional_edits(self, additional_edits: List[dict]) -> None:
-        edits = list(parse_text_edit(additional_edit) for additional_edit in additional_edits)
-        debug('applying additional edits:', edits)
-        self.view.run_command("lsp_apply_document_edit", {'changes': edits})
-        sublime.status_message('Applied additional edits for completion')
+class LspCompleteTextEditCommand(LspCompleteCommand):
+
+    def run(self, edit: sublime.Edit, **item: Any) -> None:
+        text_edit = item["textEdit"]
+        new_text = text_edit['newText']
+        edit_region = range_to_region(Range.from_lsp(text_edit['range']), self.view)
+        if item.get("insertTextFormat", InsertTextFormat.PlainText) == InsertTextFormat.Snippet:
+            self.handle_text_edit_snippet(edit, edit_region, new_text)
+        else:
+            self.handle_text_edit_plaintext(edit, edit_region, new_text)
+        self.handle_additional_edits(item)
+
+    def translated_regions(self, edit_region: sublime.Region) -> Generator[sublime.Region, None, None]:
+        selection = self.view.sel()
+        primary_cursor_position = selection[0].b
+        for region in reversed(selection):
+            # For each selection region, apply the same removal as for the "primary" region.
+            # To do that, translate, or offset, the LSP edit region into the non-"primary" regions.
+            # The concept of "primary" is our own, and there is no mention of it in the LSP spec.
+            translation = region.b - primary_cursor_position
+            translated_edit_region = sublime.Region(edit_region.a + translation, edit_region.b + translation)
+            yield translated_edit_region
```

Comment:
I wonder about performance differences with using generators as opposed to just return plain list.

I would assume it's less performant as it's basically a function call to get each item. And since we always calculate all items (we don't stop iteration early based on condition in the caller), I'm not sure it's really worth to use generators in such cases.

Reply:
I *love* generators. The user isn't writing into the text buffer with 1000's of cursors, no?